velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -782,21 +838,41 @@ std::shared_ptr<exec::VectorFunction> createDecimalFunction(
 
 VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
     udf_decimal_add,
-    decimalAddSubtractSignature(),
+    decimalAddSubtractSignature(true),
     createDecimalFunction<Addition>);
 
 VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
     udf_decimal_sub,
-    decimalAddSubtractSignature(),
+    decimalAddSubtractSignature(true),
     createDecimalFunction<Subtraction>);
 
 VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
     udf_decimal_mul,
-    decimalMultiplySignature(),
+    decimalMultiplySignature(true),
     createDecimalFunction<Multiply>);
 
 VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
     udf_decimal_div,
-    decimalDivideSignature(),
+    decimalDivideSignature(true),
+    createDecimalFunction<Divide>);
+
+VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
+    udf_decimal_add_not_allow_precision_loss,
```

Comment:
> This decimal function signature is only used in test, so register it in test scope.
Downstream project can only set the query config to enable this feature.

Looks like the declaration is only used by test code. Can we also test the false behavior via setting the query config?

Reply:
I have added

Reply:
Do we still need the declarations and registrations then? 
https://github.com/facebookincubator/velox/pull/10383/files#diff-257bd1947469dfd8494c7e19f535d4a3247b3df086287110e53740c199104073R27-R37

Reply:
Yes, it is different, with the registrations , we can test by the function name.

Reply:
I feel these declarations need to be reconsidered for below reasons.
1) they are in the runtime code but are only used by tests.
2) their signatures are generated with `allowPrecisionLoss` being false, but `createDecimalFunction`  follows the query config to decide if the function allows precision loss. Therefore, its behavior can vary according to config but its signature is fixed.

To solve above issues, maybe we need to register them as special forms. How do you think? Thanks.

Reply:
For the first question, we can make the class be public, then the signature can only be in test code.
For the second issue, for all the decimal functions, we just need signature in test code, can we support to register function without function signature?

What do you think? @mbasmanova

Reply:
I could change it like udf_concat_row, this will be a small refactor. 
```
VELOX_DECLARE_VECTOR_FUNCTION_WITH_METADATA(
    udf_concat_row,
    std::vector<std::shared_ptr<exec::FunctionSignature>>{},
    exec::VectorFunctionMetadataBuilder().defaultNullBehavior(false).build(),
    std::make_unique<RowFunction>());
```

Reply:
And it still has a problem, we need to register the decimal add name from `add` to `decimal_add` because the name should be unique in special form, can we register it as `add` for substrait plan conversion?

Reply:
We should also change here https://github.com/facebookincubator/velox/blob/main/velox/expression/ExprCompiler.cpp#L407 to break the special form name is unique, check if we can constructSpecialForm
Can we do that? @mbasmanova @rui-mo 

Which way is better?
1. current implementation;
2. special form registered as `decimal_add`, substrait function resolution unique logic;
3. special form resgistered as `add`, break special form name unique constraint.

Reply:
Refactor to current version, remove the most of the signatures, only keep input type as decimal and output type as unknown, construct `Expr` to test.

Reply:
Now the function signature requires `returnType`, but only `input type` and `function name` should specify the function, this also aligns with the sql function definition, we query by `select hash(col);` without return type.
And we don't permit the function with same argument and different return type, so I propose to change `returnType` in `FunctionSignature` to be optional, and for decimal functions, we don't need to supply returnType, the signature will be
```
  return {exec::FunctionSignatureBuilder()
              .integerVariable("a_precision")
              .integerVariable("a_scale")
              .integerVariable("b_precision")
              .integerVariable("b_scale")
              .argumentType("DECIMAL(a_precision, a_scale)")
              .argumentType("DECIMAL(b_precision, b_scale)")
              .build()};
```
@mbasmanova How about that?

Reply:
CC @zhouyuan